File: selenium-java-ui-demo/src/test/java/com/demo/saucedemo/pages/LoginPage.java
package com.demo.saucedemo.pages;

import com.demo.base.Config;
import com.demo.saucedemo.data.SauceUser;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.support.ui.WebDriverWait;

public class LoginPage extends SauceBasePage {

  private final By usernameInput = By.cssSelector("[data-test='username']");
  private final By passwordInput = By.cssSelector("[data-test='password']");
  private final By loginButton = By.cssSelector("[data-test='login-button']");
  private final By errorMessage = By.cssSelector("[data-test='error']");

  public LoginPage(WebDriver driver, WebDriverWait wait) {
    super(driver, wait);
  }

  public void goTo() {
    driver.get(Config.get("saucedemo.baseUrl") + "/");
  }

  public void login(SauceUser user) {
    driver.findElement(usernameInput).sendKeys(user.username());
    driver.findElement(passwordInput).sendKeys(user.password());
    driver.findElement(loginButton).click();
  }

  public By errorMessageLocator() {
    return errorMessage;
  }
}

File: webdriverio-ui-demo/src/websites/saucedemo/pages/LoginPage.ts
import { $ } from '@wdio/globals';
import type { ChainablePromiseElement } from 'webdriverio';
import { SauceBasePage } from './SauceBasePage.js';
import type { SauceUser } from '../data/users.js';

/**
 * Exported as a ready-made singleton (WebdriverIO's recommended POM
 * pattern, since `browser`/`$` are already global — there's no per-test
 * `page` handle to inject like in Playwright). Never `new LoginPage()` in a
 * spec; import this default export instead.
 */
class LoginPage extends SauceBasePage {
  get usernameInput(): ChainablePromiseElement {
    return $('[data-test="username"]');
  }

  get passwordInput(): ChainablePromiseElement {
    return $('[data-test="password"]');
  }

  get loginButton(): ChainablePromiseElement {
    return $('[data-test="login-button"]');
  }

  get errorMessage(): ChainablePromiseElement {
    return $('[data-test="error"]');
  }

  async goto(): Promise<void> {
    await browser.url('/');
  }

  async login(user: SauceUser): Promise<void> {
    await this.usernameInput.setValue(user.username);
    await this.passwordInput.setValue(user.password);
    await this.loginButton.click();
  }
}

export default new LoginPage();
